fe/fe-core/src/main/java/com/starrocks/sql/analyzer/SetStmtAnalyzer.java

```diff
@@ -64,22 +64,55 @@
 import org.apache.commons.lang3.StringUtils;
 
 import java.util.ArrayList;
+import java.util.HashMap;
 import java.util.List;
 
 public class SetStmtAnalyzer {
     public static void analyze(SetStmt setStmt, ConnectContext session) {
         List<SetListItem> setVars = setStmt.getSetListItems();
-        for (SetListItem var : setVars) {
-            if (var instanceof SystemVariable) {
-                analyzeSystemVariable((SystemVariable) var);
-            } else if (var instanceof UserVariable) {
-                analyzeUserVariable((UserVariable) var);
-            } else if (var instanceof SetUserPropertyVar) {
-                analyzeSetUserPropertyVar((SetUserPropertyVar) var);
-            } else if (var instanceof SetNamesVar) {
-                analyzeSetNames((SetNamesVar) var);
-            } else if (var instanceof SetPassVar) {
-                analyzeSetPassVar((SetPassVar) var, session);
+        HashMap<String, UserVariable> cloneUserVars = new HashMap<>();
+        boolean hasUserVar = setVars.stream().anyMatch(var -> var instanceof UserVariable);
+        //Record the order in which user variables appear in SQL
+        //set @a = 1, @b = @c+1, @c = 1; var b shouldn't be reanalyzed.
+        List<String> userVarSequenceInSql = new ArrayList<>();
+        if (hasUserVar) {
+            cloneUserVars.putAll(session.getUserVariables());
+        }
+        try {
+            for (SetListItem var : setVars) {
+                if (var instanceof SystemVariable) {
+                    analyzeSystemVariable((SystemVariable) var);
+                } else if (var instanceof UserVariable) {
+                    UserVariable userVar = (UserVariable) var;
+                    UserVariable userVariableExisted = session.getUserVariable(userVar.getVariable());
+                    if (userVariableExisted != null) {
+                        session.removeUserVariable(userVar.getVariable());
```

Comment:
why remove user variable here?

Reply:
During analysis, UserVariable depends on variables that may be old values.
 Under the same session, the following two SQLs are available
 set @A = 5, @b = @A + 1, @c = @b + 1;
 set @A = 6, @b = @A + 1, @c = @b + 1;
sql1 result is 5,6,7, The execution steps for sql2 are as follows
after SetStmtAnalyzer analyze the sql
var a's evaluatedExpression is IntLiteral(6)。
var b's evaluatedExpression is IntLiteral(5+1),because ConnectContext.userVariables records that the value of variable a is 5。
var c's evaluatedExpression is IntLiteral(6+1),because ConnectContext.userVariables records that the value of variable b is 6。